interfaces: add media-hub interface

alfonsosanchezbeato opened the pull request:
interfaces: add media-hub interface

alfonsosanchezbeato commented:
Apparently tests/main/interfaces-network-bind is making all checks fail. This seems completely unrelated to the PR.

stolowski commented:
@alfonsosanchezbeato FYI, I've updated this PR for the latest API changes in master.

alfonsosanchezbeato commented:
@stolowski awesome, thanks for handling this.

jdstrand commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,190 @@
+    bus=system
+    path=/com/canonical/powerd
+    interface=com.canonical.powerd,
+`
```
Can you add mediaHubPermanentSlotSeccomp with the following:
```
const mediaHubPermanentSlotSeccomp = `
bind
`
```
Your media-hub snap may be specifying 'plugs: [ network-bind ]' that is giving you this syscall, but the slots side should have it on its own to bind to the DBus well-known name.

alfonsosanchezbeato replied:
Done

jdstrand commented on interfaces/builtin/media_hub_test.go:
```diff
@@ -0,0 +1,208 @@
+	c.Assert(apparmorSpec.SecurityTags(), DeepEquals, []string{"snap.other.app"})
+	c.Assert(apparmorSpec.SnippetForTag("snap.other.app"), Not(IsNil))
+	c.Assert(apparmorSpec.SnippetForTag("snap.other.app"), testutil.Contains,
+		`#include <abstractions/dbus-session-strict>`)
```
This isn't specific to the plug side. Can you add another assert for a string that is specific to only the plug side?

jdstrand commented on interfaces/builtin/media_hub_test.go:
```diff
@@ -0,0 +1,208 @@
+	spec := &seccomp.Specification{}
+	c.Assert(spec.AddConnectedPlug(s.iface, s.plug, s.slot), IsNil)
+	c.Assert(spec.Snippets(), HasLen, 0)
+}
```
Feel free to drop this since there is no connected plug seccomp policy.

alfonsosanchezbeato commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,190 @@
+
+# DBus accesses
+#include <abstractions/dbus-session-strict>
+#include <abstractions/dbus-strict>
```
Accessing to the system bus is needed to access:

com.canonical.Unity.Screen
com.canonical.powerd

alfonsosanchezbeato replied:
Note that even although the media-hub snap has a screen-inhibit-control plug, media-hub-server crashes without this line as it tries to access powerd.

alfonsosanchezbeato commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,190 @@
+dbus (receive, send)
+    bus=system
+    path=/com/canonical/powerd
+    interface=com.canonical.powerd,
```
This interface is used to access powerd (repowerd now) in Touch. If at some time that sw is snapped, it would make sense to create its own interface. I have removed this as apparently that has no adverse effects for media-hub.

alfonsosanchezbeato commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,190 @@
+    path=/core/ubuntu/media/Service
+    peer=(label=###PLUG_SECURITY_TAGS###),
+
+dbus (receive)
```
Done

alfonsosanchezbeato commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,190 @@
+dbus (receive)
+    bus=session
+    interface=org.freedesktop.DBus.Introspectable
+    peer=(label=###PLUG_SECURITY_TAGS###),
```
Done

alfonsosanchezbeato commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,190 @@
+    bus=session
+    interface="core.ubuntu.media.Service{,.*}"
+    path=/core/ubuntu/media/Service
+    peer=(label=###PLUG_SECURITY_TAGS###),
```
That's right, there is lots of overlap in this section. I have reformatted it. We consider access to core.ubuntu.media.Service{,.*} safe, as is the client interface.

jdstrand replied:
Ok, thanks.

alfonsosanchezbeato commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,190 @@
+dbus (receive, send)
+    bus=session
+    path=/core/ubuntu/media/Service{,/**}
+    peer=(label=###SLOT_SECURITY_TAGS###),
```
Done, now this is symmetrical with the connected slot.

alfonsosanchezbeato commented:
@jdstrand thanks for the review! MP refreshed after addressing review comments. Removing

#include <abstractions/dbus-strict>

is the only comment that needs some discussion.

alfonsosanchezbeato commented:
@jdstrand, actually in the end not having  <abstractions/dbus-strict> is not an issue, as it gets included when needed (that is, via screen-inhibit-control interface)

zyga commented:
Hey @alfonsosanchezbeato can you please merge master into this branch?
@jdstrand can you please re-review this?

alfonsosanchezbeato commented:
@zyga , done.

jdstrand commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,189 @@
+
+const mediaHubPermanentSlotSecComp = `
+# Description: Allow operating as the media-hub service. This gives
+# privileged access to the system.
```
'This gives privileged access to the system can be dropped'. The 'bind' here doesn't give that.

jdstrand commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,189 @@
+    path=/core/ubuntu/media/Service
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+`
```
Please remove this blank line.

jdstrand commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,189 @@
+dbus (receive, send)
+    bus=session
+    path=/core/ubuntu/media/Service/sessions/**
+    interface="org.mpris.MediaPlayer2{,.Player,.TrackList}",
```
This lists a 'TODO'. What happened with that?

Also, this rule does not use 'peer=(label=...)'. Please add:
```
peer=(label=unconfined),
```

jdstrand commented on interfaces/builtin/media_hub.go:
```diff
@@ -0,0 +1,189 @@
+dbus (send)
+    bus=session
+    path=/core/ubuntu/media/Service/sessions/**
+    interface="org.freedesktop.DBus.Properties",
```
This rule does not use 'peer=(label=...)'. Please add:
```
peer=(label=unconfined),
```

alfonsosanchezbeato commented:
@jdstrand, thanks for the review. All comments addressed and branch refreshed. Also, I have removed the TODO line, as it is actually possible that media-hub receives mpris signals, and that is fine.

mvo5 commented:
Looks good
